### ISSUE TITLE
Restrict the use of `approve-and-merge`

### ISSUE THREAD

**Issue by bnchrch** [airbytehq/airbyte#31251]
## Problem
`approve-and-merge` use is a matter of process. Critical CI is being skipped.

This PR is the first part in a series of steps to remove approve and merge.

It focuses on restricting opting out of CI too early.

(See Original write up by @postamar below for a much better description.)

## Solution
🚧 This is not the entire solution proposed by @postamar instead its a modification based on the [slack thread](https://airbytehq-team.slack.com/archives/C02UF50V9HA/p1696626683363899) and the idea of we should apply pressure to this use case first before fully removing it

This Issue represent multiple small changes that should be made:
1. Update version qa check to ignore test paths
2. Do not run gradle check on docs
3. Update approve-and-merge to require that the following checks have passed (or skipped)
    1. Checks (gradle)
    2. Formatting
    3. Connector Tests
### Why are we not skipping connector tests for non certified connectors?
I believe that is counter to our goals. If you connector is not meeting the bare minimum of CAT test then it should not be merged

### Why are we not deleting `approve-and-merge`
Not yet! I worry that to remove it will increase late night OC pings. We should see that from the usage later
## Original
Original credit to @postamar [slack thread](https://airbytehq-team.slack.com/archives/C02UF50V9HA/p1696626683363899)

### Copy and paste of slack
**Proposal for getting rid of **`**/approve-and-merge**`** in airbytehq/airbyte***
tl;dr: right now it's only the connectors CI tests, aka airbyte-ci, that get in the way of a merge*Today I took a look at [#bot-approved-prs](https://airbytehq-team.slack.com/archives/C04928XA0LB). In September 2023, 50 PRs to airbytehq/airbyte were merged using `/approve-and-merge`. I looked at each of these PRs (see <https://docs.google.com/spreadsheets/d/1s53lfSr3nMDcNpTRcrvzfwsJ2fyPPJIn91TActZfQjc> ) and was happy to find no instances of gross abuse. People really are behaving reasonably ![:heart:](https://a.slack-edge.com/production-standard-emoji-assets/14.0/apple-medium/2764-fe0f@2x.png)Of these 50 PRs:

-   62% were connector changes, 24% were build infra changes (gradle, airbyte-ci, etc.), 14% were documentation-only changes.
-   66% used the command because CI was blocking a merge, the remaining 34% had other reasons which all boil down to "didn't feel like waiting".

Breaking down that 66% blocked-because-of-CI further:

1.  75% of the time, it's the connectors CI tests that's in the way of a merge,
2.  15% of the time it's broken formatting,
3.  10% of the time it's some other required gradle check.

Let's first consider formatting and other required gradle checks separately from connectors CI. These days these repo-wide checks behave fairly predictably, assuming that they don't fail on master there's no good reason whatsoever to merge a PR with changes which causes to fail them. The best reason I've heard so far is "can't wait for check results these take too long", which won't be valid for long as connector-ops and myself are going to bring these down from the ~18 minutes they currently take to something much less substantial before the end of Q4a. I don't see why having to wait 7 or 8 minutes after pushing is anything but a very mild annoyance, certainly far less of an annoyance than breaking the build on master is for other developers. In the special case of documentation-only changes, it should be trivial to add a filter to the gradle check which skips it if the git diff only spans `docs/**` or `**/*,md` files or whatever. Flaky tests, i.e. which fail nondeterministically, can be skipped and left to be deflaked asynchronously. Auto-formatting can be disabled, because the repo-wide checks include formatting checks.This leaves us with the connectors CI tests. If we stopped making this a required status check, then it follows that we could disable `/approve-and-merge` literally right now. We can do better, though. I haven't done an analysis of the connectors CI tests' false negatives, but I do know that we've reached a point it can be trusted for a small subset of connectors, which basically includes all certified connectors plus those we care about like connectors under active development. Rather than being able to ignore the whole connectors pipeline, let's instead maintain a whitelist of connectors for which a failure prevents merging the PR. Testing the remaining connectors will be done for advisory purposes only, which is already the way we consume those reports today in practice. This whitelist may perhaps be implicitly derived from the `metadata.yml`, who knows.To summarize, here's my proposal:

-   split the connectors pipeline into two: one for whitelisted connectors (reassign the existing "Connectors CI" status check to it) and another for the remaining connectors;
-   skip the "Gradle Check" status check for docs-only changes;
-   speed up the "Gradle Check" job, otherwise people will be annoyed;
-   remove the auto-formatting github workflow and associated status check, it will no longer be needed;
-   disable `/approve-and-merge` -- if ever anyone *really* needs to do something naughty, they can involve an admin.

It's now time we had CI that works for us instead of against us. I hope this wall of text at least convinces everyone that reads it that fixing our CI right now requires little more than formalizing our expectations on quality for the repo. For instance, we expect formatting to not be broken anyway and any java source files to at least compile and source-postgres to pass all tests, but we don't expect the same from, say, source-sftp.discussion in ![:thread:](https://a.slack-edge.com/production-standard-emoji-assets/14.0/apple-medium/1f9f5@2x.png)
